src/drivers/driver_zephyr.c

```diff
@@ -236,6 +237,444 @@ static void wpa_drv_zep_event_proc_unprot_disassoc(struct zep_drv_if_ctx *if_ctx
+	int ret;
+
+	if (phy_info->last_mode != band_info->band) {
+		mode = os_realloc_array(phy_info->modes,
+				*phy_info->num_modes + 1,
+				sizeof(*mode));
+
+		if (!mode) {
+			phy_info->failed = 1;
+			return -1;
+		}
+
+		phy_info->modes = mode;
+
+		mode = &phy_info->modes[*(phy_info->num_modes)];
+
+		os_memset(mode, 0, sizeof(*mode));
+
+		mode->mode = NUM_HOSTAPD_MODES;
+		mode->flags = HOSTAPD_MODE_FLAG_HT_INFO_KNOWN |
+			HOSTAPD_MODE_FLAG_VHT_INFO_KNOWN;
+
+		/*
+		 * Unsupported VHT MCS stream is defined as value 3, so the VHT
+		 * MCS RX/TX map must be initialized with 0xffff to mark all 8
+		 * possible streams as unsupported. This will be overridden if
+		 * driver advertises VHT support.
+		 */
+		mode->vht_mcs_set[0] = 0xff;
+		mode->vht_mcs_set[1] = 0xff;
+		mode->vht_mcs_set[4] = 0xff;
+		mode->vht_mcs_set[5] = 0xff;
+
+		*(phy_info->num_modes) += 1;
+
+		phy_info->last_mode = band_info->band;
+		phy_info->last_chan_idx = 0;
+	}
+	else
+		mode = &phy_info->modes[*(phy_info->num_modes) - 1];
+
+	phy_info_ht_capa_cfg(mode, band_info->ht_cap.wpa_supp_cap,
+			band_info->ht_cap.wpa_supp_ampdu_factor,
+			band_info->ht_cap.wpa_supp_ampdu_density,
+			&band_info->ht_cap.mcs);
+
+	phy_info_vht_capa_cfg(mode, band_info->vht_cap.wpa_supp_cap,
+			&band_info->vht_cap.vht_mcs);
+
+	ret = phy_info_freqs_cfg(phy_info, mode, band_info);
+
+	if (ret == 0)
+		ret = phy_info_rates_cfg(mode, band_info);
+
+	if (ret != 0) {
+		phy_info->failed = 1;
+		return ret;
+	}
+
+	return 0;
+}
+
+static void wpa_drv_zep_event_get_wiphy(struct zep_drv_if_ctx *if_ctx, void *band_info)
+{
+	if (!band_info) {
+		if_ctx->get_wiphy_in_progress = false;
+		return;
+	}
+
+	phy_info_band_cfg(if_ctx->phy_info_arg, band_info);
+}
+
+static int wpa_drv_register_frame(struct zep_drv_if_ctx *if_ctx,
+		u16 type, const u8 *match, size_t match_len,
+		bool multicast)
+{
+	const struct zep_wpa_supp_dev_ops *dev_ops = NULL;
+
+	dev_ops = if_ctx->dev_ctx->config;
+
+	if (!dev_ops->register_frame)
+		return -1;
+
+	return dev_ops->register_frame(if_ctx->dev_priv, type, match, match_len, false);
+}
+
+static int wpa_drv_register_action_frame(struct zep_drv_if_ctx *if_ctx,
+		const u8 *match, size_t match_len)
+{
+	u16 type = (WLAN_FC_TYPE_MGMT << 2) | (WLAN_FC_STYPE_ACTION << 4);
+
+	return wpa_drv_register_frame(if_ctx, type, match, match_len, false);
+}
+
+static int wpa_drv_mgmt_subscribe_non_ap(struct zep_drv_if_ctx *if_ctx)
+{
+	int ret = 0;
+
+	/* WNM - BSS Transition Management Request */
+	if (wpa_drv_register_action_frame(if_ctx, (u8 *)"\x0a\x07", 2) < 0)
+		ret = -1;
+
+	return ret;
+}
+
+static void wpa_drv_zep_event_mgmt_rx(struct zep_drv_if_ctx *if_ctx,
+		char *frame, int frame_len,
+		int frequency, int rx_signal_dbm)
+{
+	const struct ieee80211_mgmt *mgmt;
+
+	union wpa_event_data event;
+	u16 fc, stype;
+	int rx_freq = 0;
+
+	wpa_printf(MSG_MSGDUMP, "wpa_supp: Frame event");
+	mgmt = (const struct ieee80211_mgmt *)frame;
+
+	if (frame_len < 24) {
```

Comment:
better to use a sizeof() if we have one of a define

Reply:
Again, sticking to nl80211 logic for now since sizeof is much larger than 24.